modules/shared/src/main/scala/org/tessellation/merkletree/MerkleTree.scala

```diff
@@ -0,0 +1,128 @@
+package org.tessellation.merkletree
+
+import cats.data.NonEmptyList
+import cats.kernel.Eq
+import cats.syntax.eq._
+import cats.syntax.option._
+
+import org.tessellation.security.hash.Hash
+
+import eu.timepit.refined.cats._
+import eu.timepit.refined.types.numeric.NonNegInt
+
+case class ProofEntry(target: Hash, siblings: Either[Hash, Hash])
+
+case class Proof(entries: NonEmptyList[ProofEntry]) {
+  def verify(candidate: Hash): Boolean =
+    entries
+      .foldLeft(candidate.some.map(MerkleTree.hashLeaf)) {
+        case (Some(curr), pe) =>
+          val lsib = pe.siblings.left.getOrElse(curr)
+          val rsib = pe.siblings.getOrElse(curr)
+          val hash = MerkleTree.hashIntermediate(lsib, rsib)
+          Option.when(hash == pe.target)(hash)
+        case (None, _) => None
+      }
+      .isDefined
+}
+
+case class MerkleTree(leafCount: NonNegInt, nodes: NonEmptyList[Hash]) {
+  def getRoot: Hash =
+    nodes.last
+
+  def findPath(node: Hash): Option[Proof] =
```

Comment:
Aren't we looking for a path to a leaf rather than a node?

Reply:
when we create merkle tree, we provide `nodes`, so here is a `node` as well. I can change it to `element` (?)

Reply:
nvm. changed to `leaf`